ipykernel/kernelbase.py

```diff
@@ -170,6 +172,26 @@ def __init__(self, **kwargs):
         for msg_type in self.control_msg_types:
             self.control_handlers[msg_type] = getattr(self, msg_type)
 
+        self._stdin_msg = None
+        self._stdin_lock = threading.Lock()
+
+        if isinstance(self.stdin_socket, DummySocket):
+            # This is a test
+            self.stdin_stream = None
+            return
+        self.stdin_stream = ZMQStream(self.stdin_socket)
+
+        def handle_msg(msg):
+            idents, msg = self.session.feed_identities(msg, copy=False)
+            try:
+                msg = self.session.deserialize(msg, content=True, copy=False)
+            except Exception:
+                self.log.error("Invalid Message", exc_info=True)
+                return
+            self._stdin_msg = msg
```

Comment:
Don't we need to lock here too?

Reply:
The messages are processed by a single thread so a lock would not be useful here.